text/1193274/tex/torus.tex

```diff
@@ -0,0 +1,108 @@
+natural globally consistent way (Figure 2). Having such a labeling, it is trivial to check if 2 paths lead to the same processor. The lack of sense of direction is known to increase the message complexity of the election problem. For example in a \textit{torus} we can choose leader in $\Theta(n)$, instead of $\Theta(n\log{}n)$ \cite{mans}, when nodes have the sense of direction.
+
+\begin{figure}[h!]
+\centering
+\includegraphics[scale=0.9]{labels.png}
+\caption{Labels of the edges \cite{mans}}
+\label{fig:torus}
+\end{figure}
+
+\newpage
+
+\section{Outline of leader election algorithm}  
+The main idea behind Peterson’s algorithm for square bidirectional \textit{tori} is marking territory. In each stage, each processor tries to mark off the boundary of a square distance $d$ on a side ($d=\alpha^i$ for some constant $\alpha$). If node closes boundary, it survives to the next stage. If node's message encounters territory with lower \textit{id}. it does not survive, because boundary with lower \textit{id} purges its message. Peterson shows, that $\alpha=1.1795$ to obtain $\Theta(n)$. Peterson also suggested, that “the algorithm only
+needs to mark off a square; the orientation of the square
+is irrelevant.”\cite{peterson}
+
+\section{Getting a local sense of direction}
+
+\begin{subsection}{Main idea}
+We only need the processor to pass message in straight line or make appropriate turn. Is relatively easy to obtain. Each processor has to know the identity and the position of each processor at distance 2.   
+\end{subsection}
+
+\begin{subsection}{Algorithm}
+In a nutshell, each processor sends its identity to each of its neighbors, which forwards it to its three remaining neighbors.
+
+\begin{algorithm} 
+\caption{Bernard Mans' algorithm for local sense of direction in \textit{tori}}
+\begin{algorithmic}
+\State $\forall \ \text{link} \ r \ \text{initialize} \ H1_r:=\emptyset,\ H2_r:=\emptyset$
+\State $\forall \ \text{link} \ k \ \text{send ONE(myId) on arc k}$
+\\
+\Repeat
+    \State \Call{Receive}{}
+\Until received 16 messages
+\\
+\Procedure{Receive}{$message(type, \ id, \ link)$}
+    \If{$type = ONE$}
+        \State $H1_\text{link}:=\{id\}$
+        \State $\forall \ \text{link} \ k \neq link  \ \text{send TWO(id) on arc k}$
+    \EndIf
+    \If{$type = TWO$}
+        \State $H2_\text{link}:=H2_\text{link} \cup \{id\}$    
+    \EndIf
+\EndProcedure
+\end{algorithmic}
+\end{algorithm}
+
+\newpage
+
+After execution of this algorithm each node knows:
+\begin{itemize}
+    \item its immediate neighbours
+    \item links must be used to reach them
+    \item nodes at distance 2
+\end{itemize}
+The amount of information is enough to pass messages straight and make appropriate turns. If $H2_i \cap H2_j = \emptyset$, then links $i$ and $j$ are parallel, else they are perpendicular. (this statement is $true$ for $\sqrt{n} > 4$) 
+
```

Comment:
BTW do you address the case $n \le 2$?

Reply:
For $n \leq 2$, it is obvious, how to choose a leader, torus consists of only 1 processor ;). Cases $4 \leq n \leq 16$ are not covered in the article, so I decided I also won't cover it (parallel $H2_i, H2_j$ have non-empty intersection).